### ISSUE TITLE
[BUG] TypeError: val.toObject is not a function

### ISSUE THREAD

**Issue by simllll** [Automattic/mongoose#9518]
*bug*

**What is the current behavior?**
mongoose fails with val.toObject is not a function

 {"stack":"TypeError: val.toObject is not a function\n    at modifiedPaths (/home/simon/Dev/hokify/hokify-server/node_modules/mongoose/lib/helpers/common.js:84:17)\n    at modifiedPaths (/home/simon/Dev/hokify/hokify-server/node_modules/mongoose/lib/helpers/common.js:87:7)\n    at modifiedPaths (/home/simon/Dev/hokify/hokify-server/node_modules/mongoose/lib/helpers/common.js:87:7)\n    at modifiedPaths (/home/simon/Dev/hokify/hokify-server/node_modules/mongoose/lib/helpers/common.js:87:7)\n    at modifiedPaths (/home/simon/Dev/hokify/hokify-server/node_modules/mongoose/lib/helpers/common.js:87:7)\n    at modifiedPaths (/home/simon/Dev/hokify/hokify-server/node_modules/mongoose/lib/helpers/common.js:87:7)\n    at modifiedPaths (/home/simon/Dev/hokify/hokify-server/node_modules/mongoose/lib/helpers/update/modifiedPaths.js:24:7)\n    at _decorateUpdateWithVersionKey (/home/simon/Dev/hokify/hokify-server/node_modules/mongoose/lib/model.js:2493:24)\n    at _update (/home/simon/Dev/hokify/hokify-server/node_modules/mongoose/lib/model.js:3759:3)\n    at Function.updateOne (/home/simon/Dev/hokify/hokify-server/node_modules/mongoose/lib/model.js:3697:10)\n    at OrganizationReviewQueueRepository.insertOrUpdate (/home/simon/Dev/hokify/hokify-server/packages/organization-data/src/database/OrganizationReviewQueueRepository.ts:203:43)\n    at OrganizationChangeService.processChange (/home/simon/Dev/hokify/hokify-server/packages/organization-service/src/OrganizationChangeService.ts:40:49)\n    at processTicksAndRejections (internal/process/task_queues.js:93:5)\n    at OrganizationService.createOrganization (/home/simon/Dev/hokify/hokify-server/packages/organization-service/src/OrganizationService.ts:835:3)\n    at SignupService.postUserSignUp (/home/simon/Dev/hokify/hokify-server/packages/user-auth/src/SignupService.ts:595:26)\n    at SignupService.processLogin (/home/simon/Dev/hokify/hokify-server/packages/user-auth/src/SignupService.ts:448:4)\n    at Strategy._verify (/home/simon/Dev/hokify/hokify-server/packages/user-auth/src/authStrategies/LocalSignupStrategy.ts:219:27)","message":"val.toObject is not a function","status":500}

**If the current behavior is a bug, please provide the steps to reproduce.**
Hard to say, it just ahppened since arround mongoose 5.10.10 somewhere.

**What is the expected behavior?**
should not fail.

**What are the versions of Node.js, Mongoose and MongoDB you are using? Note that "latest" is not a version.**
node v14.15.0
mongoose 5.10.12
mongodb 4.4.1

**Comment by simllll** [comment]
clientErrorHandler err MongoError: The dollar ($) prefixed field '$isDocumentArrayElement' in 'changes..newValue.trustedUsers.0.$isDocumentArrayElement' is not valid for storage.

somehow the reason was that i passed in a mongoose object, we only use "plain" ojbects though, so I fixed that by using .toObject() on the saved object and then pass it to the next function (where the error occured) and now things work again. But shouldn't it work with a mongoose object correctly too?

**Comment by vkarpov15** [comment]
Without more information, it's hard for me to tell what's happening. All the stack trace tells me that `updateOne()` received an update with a value that Mongoose thinks looks like a Mongoose object, but isn't actually an instance of a Mongoose class.

Are you doing some sort of cloning of Mongoose objects using `Object.assign()` or the [spread operator](https://thecodebarbarian.com/object-assign-vs-object-spread.html)?

The `The dollar ($) prefixed field '$isDocumentArrayElement'` error is expected if you bypass Mongoose, Mongoose is responsible for converting Mongoose-specific objects into POJOs.

**Comment by simllll** [comment]
I believe the error was that the "mongoose" object has been part of another object, and was passed in via spread operator as you suggested. According to chagnelog we didn't touch this code though for at least 3 months, and suddenly it's not working anymore. Or at least it throws an error now, I'm not quite sure if it ever worked ;)

It's not a real bug I guess, a better error message would just be great instead of "toObject" is not a function, maybe check for typeof obj.toObject === 'function' and if it's not throw some details about the object?

**Comment by vkarpov15** [comment]
I managed to repro a somewhat related bug, but the stack trace I get is different:

```
     TypeError: obj.toObject is not a function
      at Object.clone (lib/helpers/clone.js:48:16)
      at Object.exports.mergeClone (lib/utils.js:831:25)
      at model.Query.Query._mergeUpdate (lib/query.js:3701:11)
      at _update (lib/query.js:4233:9)
      at model.Query.Query.updateOne (lib/query.js:4136:10)
      at _update (lib/model.js:3761:16)
      at Function.updateOne (lib/model.js:3697:10)
```

Below is the script:

```javascript
'use strict';

const mongoose = require('mongoose');

mongoose.set('debug', true);

run().catch(err => console.log(err));

async function run() {
  await mongoose.connect('mongodb://localhost:27017/test', {
    useNewUrlParser: true,
    useUnifiedTopology: true
  });
  
  await mongoose.connection.dropDatabase();
  
  const schema = new mongoose.Schema({
    name: String,
    children: [{ name: String }]
  });
  
  const Person = mongoose.model('Person', schema);

    const doc = await Person.create({
    name: 'Anakin',
    children: [{ name: 'Luke' }]
  });

  doc.children[0].name = 'Luke Skywalker';
  const update = { 'children.0': { ...doc.children[0] } };
  console.log(update);
  await Person.updateOne({ _id: doc._id }, update);

  console.log('Done');
}
```

Can you please modify the above script to demonstrate your issue? I added a fix for the bug in the above script that will be released with v5.10.13
